### PR TITLE
Add project(String) method to DependencyHandler

### PR DESCRIPTION
**Pull request by ghale**
This adds a new `project(String)` method to DependencyHandler to allow adding project dependencies without using actual `Project` objects.  This clears the way for us to deprecated passing a `Project` object as a dependency notation in 9.6.

This also simplifies the dependency factory infrastructure by pushing `ProjectFinder` down to the only place it's actually used (`DefaultProjectDependencyFactory`) so that we aren't passing it around through a bunch of intermediate method calls.

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by ghale** [comment]
@bot-gradle test this

File: platforms/jvm/scala/src/integTest/groovy/org/gradle/scala/ScalaPluginIntegrationTest.groovy [hunk 1]
```diff
@@ -105,7 +105,7 @@ class ScalaPluginIntegrationTest extends MultiVersionIntegrationSpec {
                 apply plugin: 'scala'
                 dependencies {
                     implementation("${scalaDependency(version.toString())}")
-                    implementation(project(":java").sourceSets.main.output)
+                    implementation(project.project(":java").sourceSets.main.output)
```
Comment (tresat) [inline platforms/jvm/scala/src/integTest/groovy/org/gradle/scala/ScalaPluginIntegrationTest.groovy:108]: 💅 `project.project`?

File: platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/ProjectDependenciesIntegrationTest.groovy [hunk 2]
```diff
@@ -114,6 +114,23 @@ class ProjectDependenciesIntegrationTest extends AbstractDependencyResolutionTes
         failure.assertHasCause("Project with path ':unknown' could not be found.")
     }
 
+    def "can declare project dependency using project path string"() {
```
Comment (tresat) [inline platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/ProjectDependenciesIntegrationTest.groovy:117]: 🤔 It might be good to have a test for `project(":")` to demo the behavior.

File: platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/BuildscriptResolutionIntegrationTest.groovy [hunk 3]
```diff
@@ -45,7 +45,7 @@ class BuildscriptResolutionIntegrationTest extends AbstractIntegrationSpec {
                 }
 
                 dependencies {
-                    classpath project(":")
+                    classpath project.project(":")
```
Comment (tresat) [inline platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/BuildscriptResolutionIntegrationTest.groovy:48]: 💅 `project.project`?

File: platforms/jvm/scala/src/integTest/groovy/org/gradle/scala/ScalaPluginIntegrationTest.groovy (hunk 1 above)
Comment (ghale) [reply platforms/jvm/scala/src/integTest/groovy/org/gradle/scala/ScalaPluginIntegrationTest.groovy:108]: This is intentional.  In this case, we're not creating a dependency on a project, we're creating a dependency on a sourceset.  So, using the implicit delegation calls `DependencyHandler.project()` which doesn't get us what we want - we have to explicitly call `Project.project()`.

File: platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/BuildscriptResolutionIntegrationTest.groovy (hunk 3 above)
Comment (ghale) [reply platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/BuildscriptResolutionIntegrationTest.groovy:48]: Also intentional.  In this case, we're testing that declaring project dependencies in a buildscript fail to resolve.  Since this is in the buildscript, you get an error when you call `DependencyHandler.project()` (i.e. you get it immediately during declaration) so we use `Project.project()` to get the dependency declared and then verify that the resolution itself fails.

They're basically edge case tests only there to pin down behavior.

File: platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/ProjectDependenciesIntegrationTest.groovy (hunk 2 above)
Comment (ghale) [reply platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/ProjectDependenciesIntegrationTest.groovy:117]: Done.
